### DISCUSSION TITLE
Simple model won't converge (static loss)

### DISCUSSION THREAD

**Discussion by tflahaul** [google/flax#2178] [Q&A] [answered]
Hey there!
I want to transition from PyTorch to JAX, but I'm having a hard time training a simple model on the MNIST dataset. I followed the given example but my loss isn't moving, as if the gradients weren't applying to the model parameters.

Can you tell what I'm doing wrong?

```python
def create_train_state(key):
	net = ConvMixer(12, 2, num_classes=10)
	parameters = net.init(key, jnp.ones((1, 28, 28, 1)))
	optimizer = optax.sgd(config.LEARNING_RATE, momentum=config.MOMENTUM)
	return train_state.TrainState.create(
		apply_fn=net.apply,
		params=parameters.get('params'),
		tx=optimizer)

def apply_model(parameters, images, targets):
	y_hat = ConvMixer(12, 2, num_classes=10).apply({'params': parameters}, images)
	return jnp.mean(optax.softmax_cross_entropy(y_hat, jax.nn.one_hot(targets, 10)))

def main() -> None:
	dataset = MNIST('mnist.csv')
	state = create_train_state(jax.random.PRNGKey(42))
	grad_fn = jax.value_and_grad(apply_model)
	for epoch in range(1, config.MAX_ITER + 1):
		running_loss = 0
		for images, targets in batch_generator(dataset, config.BATCH_SIZE):
			loss, gradients = grad_fn(state.params, images, targets)
			state.apply_gradients(grads=gradients)
			running_loss = running_loss + loss
		print(f'epoch {epoch:>2d}/{config.MAX_ITER:<2d}| loss={running_loss:.4f}')
```

**Comment by marcvanzee** [comment, marked as answer]
`state` is an immutable datastructure and cannot be modified in place. So `state.apply_gradients(grads=gradients)` returns the new state, which means you should replace that line with:

```py
state = state.apply_gradients(grads=gradients)
```

**Reply by tflahaul** [reply]
That was the issue, thanks a lot!
